docs: external api development from localhost via tunneling (#2650)
commercetools/merchant-center-application-kit: website/src/content/concepts/integrate-with-your-own-api.mdx

```console
ngrok http <desired_port>
```
where `<desired_port>` is the port bound to your server application (for example `6000`).

The exposed external URL of the created tunnel:
![ngrok agent](/images/ngrok.png) 
can be further passed to the session middleware verifying the JWT audience claim.
```js highlightLines="10"
const express = require("express");
const {
  createSessionMiddleware,
  CLOUD_IDENTIFIERS,
} = require("@commercetools-backend/express");

const app = express();
app.use(
  createSessionMiddleware({
    audience: "https://33ea-87-183-162-9.eu.ngrok.io",
    issuer: CLOUD_IDENTIFIERS.GCP_EU,
  })
);
app.post("/", (req, res) => {
  res.send("It works 🙌");
});

app.listen(6000, () => {
  console.log("Running on port 6000.");
});

module.exports = app;
```

To verify the tunneling service, configure the additonal environment variable `echoServerApiUrl` in your Custom Application:
```js title="custom-application-config.mjs" highlightLines="4"
const config = {
  // ...
  additionalEnv: {
    echoServerApiUrl: "https://33ea-87-183-162-9.eu.ngrok.io",
  },
```
and use it in requests sent to the `/proxy/forward-to` endpoint:

> ```jsx title="echo-server.jsx" highlightLines="8-10,12,19"

Reviewer: This code snippet can also be omitted I think, as the usage of `/proxy/forward-to` is already documented elsewhere.

The important thing here is just to communicate to use a secure tunnel instead of the production URL, the logic in the code then stays the same.

const EchoServer = () => {
  // ...
  const dispatch = useAsyncDispatch();
  const echoServerApiUrl = useApplicationContext(
    (context) => context.environment.echoServerApiUrl
  );
  const handleSendRequest = useCallback(() => {
    const forwardToUrl = new URL(echoServerApiUrl);
    // ...

    async function ping() {
      try {
        const result = await dispatch(
          actions.forwardTo.post({
            uri: forwardToUrl.toString(),
            // ...
          })
        );
        // ...
      } catch (error) {
        // ...
      }
    }
    // ...
    ping();
  }, [
    dispatch,
    echoServerApiUrl,
    // ...
  ]);
}
```
<Info>

The entire implementation can be found under the following [link](https://github.com/commercetools/merchant-center-application-kit/blob/main/playground/src/components/echo-server/echo-server.jsx)

</Info>

In result, HTTP POST requests with a `x-forward-to` header, for example:
```
x-forward-to: https://33ea-87-183-162-9.eu.ngrok.io/?hello=world
```
should result in a response:
```
It works 🙌
```
<Info>

The details of all requests tunneled via `ngrok` can be investigated in the web application running at `http://localhost:4040/`

> </Info>

Author: I rephrased it a little and added the missing info that the monitoring app actually starts alongside the agent 😅 However, if you find that not needed at all, please let me know

</Info>

# Local development using a secure tunnel

Developing a Custom Application locally, together with an external API, has some limitations.
For example, using the `/proxy/forward-to` endpoint of the Merchant Center API Gateway with a target URL pointing to `http` is not allowed. Only `https` targets are allowed.

The most straighforward solution is using the production URL of the external API, which may be inconvenient in the development processs.

> Therefore, some kind of secure tunnel solution can be used to develop both Custom Application and external API together, locally.

Reviewer: ```suggestion
Therefore, some kind of secure tunnel solution can be used to develop both Custom Application and external API together, locally.

Using a secure tunnel ensures that the URL used to connect to the API via the `/proxy/forward-to` endpoint is a valid remote URL with `https` protocol.
```

## ngrok

One of the most popular secure tunnel solutions is [ngrok](https://ngrok.com/), which is a reverse proxy allowing to establish an encrypted tunnel between the `ngrok` agent application running locally and the upstream service.

Please follow the [instructions](https://ngrok.com/download) to sign up and download the agent application. Once installed, start the agent application by running:

```console
ngrok http <desired_port>
```

where `<desired_port>` is the port bound to your server application (for example `6000`).

The exposed external URL of the created tunnel:

![ngrok agent](/images/ngrok.png) 

must be provided to the session middleware verifying the JWT audience claim.

```js highlightLines="10"
const express = require("express");
const {
  createSessionMiddleware,
  CLOUD_IDENTIFIERS,
} = require("@commercetools-backend/express");

const app = express();
app.use(
  createSessionMiddleware({
    audience: "https://33ea-87-183-162-9.eu.ngrok.io",
    issuer: CLOUD_IDENTIFIERS.GCP_EU,
  })
);
app.post("/echo", (req, res) => {
  res.send("It works 🙌");
});

app.listen(6000, () => {
  console.log("Running on port 6000.");
});

module.exports = app;
```

> In your Custom Application, we recommend configuring the external API URL through an environment variable, referenced in the application config through a [variable placeholder](/api-reference/application-config#using-variable-placeholders).
> That would allow to keep the value for the development environment and for the production environment separated.

Reviewer: ````suggestion
We recommend configuring the external API URL using environment variables to be able to define different values for development (using a secure tunnel) and production.
In the Custom Application config you can specify the external API URL as an [`additionalEnv `](/api-reference/application-config#additionalenv) property using a [variable placeholder](/api-reference/application-config#using-variable-placeholders).

```json
{
  "additionalEnv": {
    "externalApiUrl": "${env:EXTERNAL_API_URL}"
  }
}
```

In the application code you can then access the value using the [application context](/api-reference/commercetools-frontend-application-shell-connectors#custom-user-properties).
````

In the development environment configured with the external API URL exposed from `ngrok`, HTTP POST requests to `/proxy/forward-to` including the header:

```
x-forward-to: https://33ea-87-183-162-9.eu.ngrok.io/echo
```

should result in a response:

```
It works 🙌

> ```

Reviewer: I would omit this

<Info>

`ngrok` agent application starts along with a monitoring web application running at `http://localhost:4040/`, which can be helpful in the process of debugging of the tunnelled HTTP requests.

</Info>

## Cloudflare Tunnel

As an alternative, you might want to use the [Cloudflare Tunnel](https://developers.cloudflare.com/pages/how-to/preview-with-cloudflare-tunnel/) service.
After installing the `cloudflared` application and running it:

```console
cloudflared tunnel --url http://localhost:<desired_port>
```

The server application can be reached via the provided URL:
![cloudflared](/images/cloudflared.png) 

> 

Reviewer: ```suggestion

The same setup as `ngrok` applies here as well.
```
